src/analysis/souffle/may_will.dl

```diff
@@ -37,8 +37,23 @@ disallowedUsage(dataConsumer, usage, owner, tag) :-
 .decl allUsagesAllowed(dummy_var: symbol)
 
 .decl isUsage(usage: Usage)
+// DO NOT SUBMIT
+.output isUsage(IO=stdout)
 
-allUsagesAllowed("dummy_var") :- !disallowedUsage(dataConsumer, usage, owner, tag),
-    isPrincipal(dataConsumer), isUsage(usage), isTag(tag), ownsTag(owner, tag).
+isPrincipal(speaker), isPrincipal(actor), isUsage(usage), isTag(tag) :-
+  saysMay(speaker, actor, usage, tag).
+
+isPrincipal(speaker), isUsage(usage), isAccessPath(ap) :- saysWill(speaker, usage, ap).
+
+allUsagesAllowed("dummy_var") :- count : disallowedUsage(_, _, _, _) = 0.
+
+// TODO: Find a better way to do this.
+// Catch may-will issues as if they were a check.
+checkAndResult("may_will", "FAIL") :- !allUsagesAllowed("dummy_var").
+
+// We phrase this as depending on saysMay so that tests that do not have "may"
+// or "will" involved will not be considered as having had a check of this form
+// at all. Again, we should find a better way to do this.
+checkAndResult("may_will", "PASS") :- allUsagesAllowed("dummy_var").
```

Comment:
Why was the earlier version not sufficient?

Reply:
There are a few aspects to why I made this change.

1. Having the rule populating `checkAndResult("may_will"...)` occur in the boilerplate wrapper code produced by the IR translator is at odds with the goal of this PR, to make the IR translator only produce base facts. The changes of this PR make the IR translator's boilerplate contain only `.decl`s, base facts, and a few rules that are obviously aliasing (eg, `saysMay(...) :- says_may(...).`).

2. Initially, the rule only would add an entry into `testFails`; that's great if you are looking for the absence of failures, but it precludes performing other checks, like making sure that the number of `PASS`s is equal to the number of tests. This can cause confusing brittleness in checking test results. The new recipe always populates the `"may_will"` check as passing or failing.

3. The `testFails` rule must be populated by a custom rule; it is not set up to be populated by `check` facts. The new formulation can work with `check` facts.